File: daily/1014/solution.py
# -*- coding: utf-8 -*-
"""
Created on Wed Jun 17 14:12:37 2020

@author: Jianhui_Wang
执行用时 : 604 ms, 在所有 Python3 提交中击败了65.79%的用户
内存消耗 :19 MB, 在所有 Python3 提交中击败了9.09%的用户
"""

class Solution:
    def maxScoreSightseeingPair(self, A: List[int]) -> int:
        max_value = 0
        current = A[0]
        for j in range(1, len(A)):
            max_value = max(max_value, current + A[j] - j)
            current = max(current, A[j] + j)   
        return max_value

File: daily/1014/solution.cpp
//类比一个数组求最大的两数之和，关键点在于写成 (A[i]+i)+(A[j]-j) 的形式

class Solution {
public:
    int maxScoreSightseeingPair(vector<int>& A) {
        int max_value = 0;
        int current = A[0];
        for(int j = 1; j < A.size(); ++j){
            max_value = max(max_value, current + A[j] - j);
            current = max(current, A[j] + j);          
        }
        return max_value;
    }
};